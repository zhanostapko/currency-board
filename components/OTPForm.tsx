"use client";
import { useActionState, useEffect, useState } from "react";
import Input from "./Shared/Input";
import { verifyOtpAction } from "@/app/actions/verifyOtp";
import SubmitButton from "./SubmitButton";
import Spinner from "./Spinner";

type OTPFormProps = {
  email: string | null;
  otpCode: number | null;
  disableValidation?: boolean;
};

const OTPForm = ({
  email,
  otpCode,
  disableValidation = false,
}: OTPFormProps) => {
  const [state, formAction] = useActionState(verifyOtpAction, {
    error: null,
    step: "otp",
  });

  const [showOTP, setShowOTP] = useState(false);

  useEffect(() => {
    const timer = setTimeout(() => {
      setShowOTP(true);
    }, 3000);

    return () => clearTimeout(timer);
  }, []);

  return (
    <form
      noValidate={disableValidation}
      autoComplete="off"
      className="flex flex-col gap-4"
      action={formAction}
    >
      <p className="text-gray-500 text-sm">Enter the OTP sent to {email}</p>
      <Input
        id="otpCode"
        label="OTP Code"
        type="text"
        name="enteredOtp"
        required
        maxLength={6}
      />
      {state.error && <p className="text-red-500 text-xs">{state.error}</p>}
      <SubmitButton title="Verify code" />
      {/* For testing */}
      {showOTP && otpCode ? (
        <div className="mt-10 text-gray-500 text-sm transition-opacity duration-500 ease-in-out">
          Test code: <span className="font-bold">{otpCode}</span>
        </div>
      ) : (
        <Spinner size={24} />
      )}
      <input type="hidden" name="serverOtp" value={otpCode ?? ""} />
    </form>
  );
};

export default OTPForm;
