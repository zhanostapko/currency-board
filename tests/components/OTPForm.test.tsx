import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import OTPForm from "../../components/OTPForm";

jest.useFakeTimers();

jest.mock("../../app/actions/verifyOtp.ts", () => ({
  verifyOtpAction: jest.fn().mockImplementation(() => ({
    error: "Invalid OTP",
    step: "otp",
  })),
}));

describe("OTPForm Component", () => {
  const mockEmail = "test@example.com";
  const mockOtp = 123456;

  test("renders correctly with email", () => {
    const { asFragment } = render(
      <OTPForm email={mockEmail} otpCode={mockOtp} />
    );

    expect(
      screen.getByText(`Enter the OTP sent to ${mockEmail}`)
    ).toBeInTheDocument();
    expect(screen.getByLabelText("OTP Code")).toBeInTheDocument();
    expect(screen.getByText("Verify code")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("allows entering OTP and submits form", async () => {
    const { asFragment } = render(
      <OTPForm email={mockEmail} otpCode={mockOtp} />
    );

    const otpInput = screen.getByLabelText("OTP Code");
    fireEvent.change(otpInput, { target: { value: "123456" } });

    expect(otpInput).toHaveValue("123456");

    fireEvent.click(screen.getByText("Verify code"));
    expect(asFragment()).toMatchSnapshot();
  });

  test("shows error message when OTP is incorrect", async () => {
    const { asFragment } = render(
      <OTPForm email={mockEmail} otpCode={123456} />
    );

    const otpInput = screen.getByLabelText("OTP Code");
    fireEvent.change(otpInput, { target: { value: "000000" } });
    fireEvent.click(screen.getByText("Verify code"));

    await waitFor(() => {
      expect(screen.getByText("Invalid OTP")).toBeInTheDocument();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  test("displays test OTP after 3 seconds", async () => {
    const { asFragment } = render(
      <OTPForm email={mockEmail} otpCode={mockOtp} />
    );

    expect(screen.queryByTestId("spinner")).toBeInTheDocument();

    jest.advanceTimersByTime(3000);

    await waitFor(() => {
      expect(screen.queryByTestId("spinner")).not.toBeInTheDocument();
      expect(screen.getAllByText(/Test code:/i).length).toBeGreaterThan(0);
      expect(screen.getAllByText(/123456/i).length).toBeGreaterThan(0);
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
