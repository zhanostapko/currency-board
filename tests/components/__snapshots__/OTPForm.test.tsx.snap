// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OTPForm Component allows entering OTP and submits form 1`] = `
<DocumentFragment>
  <form
    action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
    autocomplete="off"
    class="flex flex-col gap-4"
  >
    <p
      class="text-gray-500 text-sm"
    >
      Enter the OTP sent to test@example.com
    </p>
    <label
      class="text-gray-400 text-left text-sm font-medium"
      for="otpCode"
    >
      OTP Code
    </label>
    <input
      class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
      id="otpCode"
      maxlength="6"
      name="enteredOtp"
      required=""
      type="text"
    />
    <p
      class="text-red-500 text-xs"
    >
      Invalid OTP
    </p>
    <button
      class="bg-blue-600 text-white py-3 rounded-lg text-center hover:bg-blue-700 transition"
      type="submit"
    >
      Verify code
    </button>
    <div
      class="flex justify-center items-center "
      data-testid="spinner"
    >
      <div
        class=" border-4 border-blue-500 border-t-transparent rounded-full animate-spin"
        style="width: 24px; height: 24px; border-width: 3px;"
      />
    </div>
    <input
      name="serverOtp"
      type="hidden"
      value="123456"
    />
  </form>
</DocumentFragment>
`;

exports[`OTPForm Component displays test OTP after 3 seconds 1`] = `
<DocumentFragment>
  <form
    action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
    autocomplete="off"
    class="flex flex-col gap-4"
  >
    <p
      class="text-gray-500 text-sm"
    >
      Enter the OTP sent to test@example.com
    </p>
    <label
      class="text-gray-400 text-left text-sm font-medium"
      for="otpCode"
    >
      OTP Code
    </label>
    <input
      class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
      id="otpCode"
      maxlength="6"
      name="enteredOtp"
      required=""
      type="text"
    />
    <button
      class="bg-blue-600 text-white py-3 rounded-lg text-center hover:bg-blue-700 transition"
      type="submit"
    >
      Verify code
    </button>
    <div
      class="mt-10 text-gray-500 text-sm transition-opacity duration-500 ease-in-out"
    >
      Test code: 
      <span
        class="font-bold"
      >
        123456
      </span>
    </div>
    <input
      name="serverOtp"
      type="hidden"
      value="123456"
    />
  </form>
</DocumentFragment>
`;

exports[`OTPForm Component renders correctly with email 1`] = `
<DocumentFragment>
  <form
    action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
    autocomplete="off"
    class="flex flex-col gap-4"
  >
    <p
      class="text-gray-500 text-sm"
    >
      Enter the OTP sent to test@example.com
    </p>
    <label
      class="text-gray-400 text-left text-sm font-medium"
      for="otpCode"
    >
      OTP Code
    </label>
    <input
      class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
      id="otpCode"
      maxlength="6"
      name="enteredOtp"
      required=""
      type="text"
    />
    <button
      class="bg-blue-600 text-white py-3 rounded-lg text-center hover:bg-blue-700 transition"
      type="submit"
    >
      Verify code
    </button>
    <div
      class="flex justify-center items-center "
      data-testid="spinner"
    >
      <div
        class=" border-4 border-blue-500 border-t-transparent rounded-full animate-spin"
        style="width: 24px; height: 24px; border-width: 3px;"
      />
    </div>
    <input
      name="serverOtp"
      type="hidden"
      value="123456"
    />
  </form>
</DocumentFragment>
`;

exports[`OTPForm Component shows error message when OTP is incorrect 1`] = `
<DocumentFragment>
  <form
    action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
    autocomplete="off"
    class="flex flex-col gap-4"
  >
    <p
      class="text-gray-500 text-sm"
    >
      Enter the OTP sent to test@example.com
    </p>
    <label
      class="text-gray-400 text-left text-sm font-medium"
      for="otpCode"
    >
      OTP Code
    </label>
    <input
      class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
      id="otpCode"
      maxlength="6"
      name="enteredOtp"
      required=""
      type="text"
    />
    <p
      class="text-red-500 text-xs"
    >
      Invalid OTP
    </p>
    <button
      class="bg-blue-600 text-white py-3 rounded-lg text-center hover:bg-blue-700 transition"
      type="submit"
    >
      Verify code
    </button>
    <div
      class="flex justify-center items-center "
      data-testid="spinner"
    >
      <div
        class=" border-4 border-blue-500 border-t-transparent rounded-full animate-spin"
        style="width: 24px; height: 24px; border-width: 3px;"
      />
    </div>
    <input
      name="serverOtp"
      type="hidden"
      value="123456"
    />
  </form>
</DocumentFragment>
`;
